Create `subgraph fetch` command

JakeDawkins opened the pull request:
This PR adds the `rover subgraph fetch` command to fetch SDL of a federated graph's implementing service from studio

Fun facts
- This command errors when being run on a non-federated graph (yay, readable errors!)
- This command errors when `--service-name` isn't found, and responds with a list of available services to query (woo)

JakeDawkins committed:
skeleton for subgraph fetch client code

JakeDawkins committed:
add subgraph fetch command

EverlastingBugstopper commented on crates/rover-client/src/query/partial/fetch.rs:
```diff
@@ -0,0 +1,187 @@
+            Ok(services.services)
+        },
+        _ => {
+            unreachable!("This case shouldn't be possible, since the operation doesn't select it!")
```
what does this mean?

JakeDawkins replied:
As I was writing my response for this comment, I think I realized this case actually _was_ potentially possible. So I'll change this to a regular error I think. But the idea was that we select `... on FederatedImplementingServices` in the operation, where it's possible to have federated services or regular ones. So we're not actually selecting any fields on non-implementing services. But I think since we're still technically selecting `__typename` on all services, there is a potential path that could lead us to that other branch of the enum

(I realize none of this probably makes sense. I'm not sure how to explain this is text very well haha)

JakeDawkins committed:
add error case for wrong graph type

EverlastingBugstopper commented on crates/rover-client/src/query/partial/fetch.rs:
```diff
@@ -43,17 +44,23 @@ fn get_services_from_response_data(
-        None => Err(RoverClientError::HandleResponse {
-            msg: "There are no implementing services in this graph".to_string(),
+        None => Err(RoverClientError::ExpectedFederatedGraph {
+            graph_name: service_name.to_string(),
```
hmmmm - the mixing of graph and service here is quite confusing

JakeDawkins replied:
Oooh good catch. That's not even the right value!

JakeDawkins committed:
update comments on future unreachable case

JakeDawkins committed:
update logs for subgraph fetch

JakeDawkins committed:
fix incorrect error usage

JakeDawkins committed:
thank you clippy

JakeDawkins committed:
Update error message & fn types

- update error message to use "graph" rather than "service"
- update util fn in rover-client/fetch to use `&str` instead of `String` to follow convention

lrlna commented on crates/rover-client/src/query/partial/fetch.rs:
```diff
@@ -0,0 +1,196 @@
+/// Snake case of this name is the mod name. i.e. fetch_subgraph_query
+pub struct FetchSubgraphQuery;
+
+/// The main function to be used from this module. This function fetches a
```
I would drop the first sentence of this doc comment. When you view this on docs.rs, it's a bit hard to figure out what `this module` is so it reads like a developer comment rather than a public facing API comment.

JakeDawkins replied:
Great point!

lrlna commented on crates/rover-client/src/query/partial/fetch.rs:
```diff
@@ -0,0 +1,196 @@
+) -> Result<String, RoverClientError> {
+    let graph_name = variables.graph_id.clone();
+    let response_data = client.post::<FetchSubgraphQuery>(variables)?;
+    let services = get_services_from_response_data(response_data, &graph_name)?;
```
I am noticing that you pass a `&str` to the `get_services_from_response_data` that then gets converted to `to_string()`. How do you feel about changing the parameters of `get_services_from_response_data` to accept an owned value right away?

lrlna replied:
Ah I see, that's not possible since our RoverClientError accepts only an owned string. Ignore the above comment then. What you have works!

lrlna commented on crates/rover-client/src/query/partial/mod.rs:
```diff
@@ -3,3 +3,6 @@ pub mod push;
 pub mod delete;
+
+/// "subgraph fetch" command execution
+pub mod fetch;
```
a comment not regarding this particular line, but the directory it's in. I think the path to this file should be `crates/rover-client/src/query/subgraph/mod.rs`. `partial` was changed to `subgraph` in Rover's main directory, and should probably be changed here as well.

lrlna commented on crates/rover-client/src/query/partial/fetch.rs:
```diff
@@ -0,0 +1,196 @@
+}
+
+#[cfg(test)]
+mod tests {
```
Lots of good tests! ✨  Thank you!

lrlna reviewed (approved):
Looks good @JakeDawkins! A few minor comments, all non-blocking.

JakeDawkins committed:
fix rebase module paths

JakeDawkins committed:
update docstring
